### ISSUE TITLE
How to support a new framework in clearml

### ISSUE THREAD

**Issue by FateScript** [allegroai/clearml#451]
Hey, thanks for your wonderful work. I want to support a new framework in clearml but I didn't find any docs.

I wonder what else should I do except writing [bind](https://github.com/allegroai/clearml/blob/master/clearml/binding/frameworks/pytorch_bind.py) of my own framework ? 
Is their any docs for this?

**Comment by bmartinn** [comment]
Thank you @FateScript ! Yes please contribute, PRs are more than welcomed :heart: 

>  I want to support a new framework in clearml but I didn't find any docs.

Yes this is actually not that trivial, hence missing docs (actually explaining how to monkey patch is quite complex :) )

Usually there are two things to patch, the first is the model load/store, the second is metric reporting.
For model patching you can check the simple `lightgbm` bind [here](https://github.com/allegroai/clearml/blob/master/clearml/binding/frameworks/lightgbm_bind.py)
For metric reporting, most frameworks these days use Tensorboard, that is already supported (i.e. all reporting are logged), if this is not the case then you can see the `xgboost` callback example [here](https://github.com/allegroai/clearml/blob/889d2373988a0d6630703cc1c865e09e58f8f981/clearml/binding/frameworks/xgboost_bind.py#L134)

Which framework are we talking about here ?
How can I help? (if you can specify the framework I might be able to suggest the easiest way to bind it)
wdyt?

**Comment by FateScript** [comment]
Thanks for your prompt reply @bmartinn . We are talking about [MegEngine](https://github.com/MegEngine/MegEngine) here. I would be appreciated if you could offer more help.

BTW, should I do anything for distributed training?

**Comment by bmartinn** [comment]
@FateScript could you point me to a simple toy example with MegEngine?
The requirements for this toy example:
- train light weight model (like mnist classification)
- produce a model/weights file
- output metrics (any metric basically)

**Comment by FateScript** [comment]
Hello @bmartinn , the following code snippet staisfy your requirments with megengine version 1.5.0, please have a try.
As for output metric, the snippet use megengine tensor. Let me down if you have any question.

```python
from megengine.data.dataset import MNIST
import megengine.module as M
import megengine.functional as F

import numpy as np

import megengine as mge
from megengine.optimizer import SGD
from megengine.autodiff import GradManager
from megengine.data import DataLoader
from megengine.data.transform import ToMode, Pad, Normalize, Compose
from megengine.data.sampler import RandomSampler

class Net(M.Module):
    def __init__(self):
        super().__init__()
        self.conv0 = M.Conv2d(1, 20, kernel_size=5, bias=False)
        self.bn0 = M.BatchNorm2d(20)
        self.relu0 = M.ReLU()
        self.pool0 = M.MaxPool2d(2)
        self.conv1 = M.Conv2d(20, 20, kernel_size=5, bias=False)
        self.bn1 = M.BatchNorm2d(20)
        self.relu1 = M.ReLU()
        self.pool1 = M.MaxPool2d(2)
        self.fc0 = M.Linear(500, 64, bias=True)
        self.relu2 = M.ReLU()
        self.fc1 = M.Linear(64, 10, bias=True)

    def forward(self, x):
        x = self.conv0(x)
        x = self.bn0(x)
        x = self.relu0(x)
        x = self.pool0(x)
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu1(x)
        x = self.pool1(x)
        x = F.flatten(x, 1)
        x = self.fc0(x)
        x = self.relu2(x)
        x = self.fc1(x)
        return x


def build_dataloader():
    # change `root` if you want save dataset in another fold
    train_dataset = MNIST(root="./dataset/MNIST", train=True, download=True)
    dataloader = DataLoader(
        train_dataset,
        transform=Compose([
            Normalize(mean=0.1307*255, std=0.3081*255),
            Pad(2),
            ToMode('CHW'),
        ]),
        sampler=RandomSampler(dataset=train_dataset, batch_size=64),
    )
    return dataloader


def train(dataloader):
    net = Net()
    net.train()
    optimizer = SGD(net.parameters(), lr=0.01, momentum=0.9, weight_decay=5e-4)
    gm = GradManager().attach(net.parameters())

    total_epochs = 10
    for epoch in range(total_epochs):
        total_loss = 0
        for step, (batch_data, batch_label) in enumerate(dataloader):
            batch_label = batch_label.astype(np.int32)
            data, label = mge.tensor(batch_data), mge.tensor(batch_label)
            with gm:
                pred = net(data)
                loss = F.loss.cross_entropy(pred, label)
                gm.backward(loss)
            optimizer.step().clear_grad()

            total_loss += loss
        print("epoch: {}, loss {}".format(epoch, total_loss/len(dataloader)))

    mge.save(net.state_dict(), 'mnist_net.mge')


def load_net():
    net = Net()
    state_dict = mge.load('mnist_net.mge')
    net.load_state_dict(state_dict)
    return net


if __name__ == "__main__":
    dataloader = build_dataloader()
    train(dataloader)
    print("load net")
    load_net()
```

**Comment by xxr3376** [comment]
You may refer to the MegEngine API doc [here](https://megengine.org.cn/doc/stable/en/reference/api/megengine.load.html), sadly we haven't translated the user guides to English yet.

**Comment by bmartinn** [comment]
@FateScript , from the example it seems that if a user wants to add some metrics to `megengine`, they need to do so manually (i.e. print, TensorBoard etc.) there is no built in auto logging of loss or something of that nature, is that accurate ?

**Comment by FateScript** [comment]
> @FateScript , from the example it seems that if a user wants to add some metrics to `megengine`, they need to do so manually (i.e. print, TensorBoard etc.) there is no built in auto logging of loss or something of that nature, is that accurate ?

Yes, for most users of `megengine`, they usually add metrics with tensorboard. If any users want to do so with clearml.Logger, I will help to contribute then.

**Comment by bmartinn** [comment]
PR with the new feature merged, closing this issue :)
